File: ruby/run-length-encoding/run_length_encoding.rb
module RunLengthEncoding
  def self.encode(plaintext)  = encode_runs(plaintext).map { |run| encode_run(run) }.join
  def self.decode(ciphertext) = decode_runs(ciphertext).map { |run| decode_run(run) }.join

  private
  def self.encode_run((letter, count)) = "#{count if count > 1}#{letter}"
  def self.decode_run((letter, count)) = "#{letter}" * count

  def self.encode_runs(plaintext)
    plaintext.scan(/(.)(\1*)/).map {|letter, repetitions| [letter, repetitions.size + 1]}.to_a
  end

  def self.decode_runs(ciphertext)
    ciphertext.scan(/(\d*)(.)/).map {|count, letter| [letter, [count.to_i, 1].max]}.to_a
  end
end

File: r/run-length-encoding/run-length-encoding.R
library(stringr)

encode <- function(plaintext) {
  runs <- str_match_all(plaintext, "(.)\\1*")[[1]]
  apply(runs, 1, encode_run) |> paste0(collapse = "")
}

encode_run <- function(run) {
  num_chars <- nchar(run[1])
  prefix <- ifelse(num_chars == 1, "", num_chars |> as.character())
  paste(prefix, run[2], sep = "")
}

decode <- function(ciphertext) {
  runs <- str_match_all(ciphertext, "(\\d+)?(.)")[[1]]
  apply(runs, 1, decode_run) |> paste0(collapse = "")
}

decode_run <- function(run) {
  count <- ifelse(is.na(run[2]), 1, run[2] |> as.numeric())
  str_dup(run[3], count)
}
